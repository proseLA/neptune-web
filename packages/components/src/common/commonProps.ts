export type CommonProps = {
  /**
   * Space-separated list of the case-sensitive CSS classes,
   * most would be applied to outermost element inside the component
   *
   * @see https://github.com/transferwise/neptune-web/blob/main/rfc/0001-always-pass-classname.md
   */
  className?: string;
};

export type AriaLabelProperty = {
  'aria-label'?: string;
};

export type LinkProps = {
  href: string;
  target?: '_blank' | '_self' | '_parent' | '_top';
};

export type Heading = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export type OnClickHandler = {
  onClick: (event: Event) => void;
};
