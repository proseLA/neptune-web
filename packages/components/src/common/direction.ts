export enum Direction {
  RTL = 'rtl',
  LTR = 'ltr',
  AUTO = 'auto',
}

export type LayoutDirection = 'rtl' | 'ltr' | 'auto';
