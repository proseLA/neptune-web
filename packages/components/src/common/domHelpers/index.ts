import { SyntheticEvent } from 'react';

export function isServerSide(): boolean {
  return typeof window === 'undefined';
}

export function isBrowser(): boolean {
  return !isServerSide();
}

export function stopPropagation(event: SyntheticEvent) {
  event.stopPropagation();
  event.preventDefault();
  if (event.nativeEvent && event.nativeEvent.stopImmediatePropagation) {
    event.nativeEvent.stopImmediatePropagation();
  }
}

export {
  addClickClassToDocumentOnIos,
  removeClickClassFromDocumentOnIos,
} from './documentIosClick';
