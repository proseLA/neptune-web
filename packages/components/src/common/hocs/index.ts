export { default as withTheme } from './theme';
