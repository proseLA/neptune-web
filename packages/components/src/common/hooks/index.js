export { default as useClientWidth } from './useClientWidth';
export { default as useConditionalListener } from './useConditionalListener';
export { default as useDirection } from './useDirection';
export { default as useHasIntersected } from './useHasIntersected';
export { default as useLayout } from './useLayout';
export { default as useTheme } from './useTheme';
