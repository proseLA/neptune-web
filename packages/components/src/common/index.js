export { default as HistoryNavigator } from './historyNavigator';
export * from './domHelpers';
export { addNoScrollBodyClass, removeNoScrollBodyClass } from './DOMOperations';
export * from './theme';
export * from './direction';
// Props
export { Priority, ControlType } from './propsValues/control';
export { Breakpoint } from './propsValues/breakpoint';
export * from './propsValues/size';
export * from './propsValues/width';
export { Type } from './propsValues/type';
export { DateMode } from './propsValues/dateMode';
export { MonthFormat } from './propsValues/monthFormat';
export * from './propsValues/position';
export { Layout } from './propsValues/layouts';
export { Status } from './propsValues/status';
export { Sentiment } from './propsValues/sentiment';
export { ProfileType } from './propsValues/profileType';
export * from './propsValues/scroll';
export { MarkdownNodeType } from './propsValues/markdownNodeType';
export { FileType } from './fileType';
export { Key } from './key';
export * from './locale';
export * from './commonProps';
