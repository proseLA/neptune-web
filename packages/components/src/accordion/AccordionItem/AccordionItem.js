import classNames from 'classnames';
import PropTypes from 'prop-types';
import { cloneElement } from 'react';

import Chevron from '../../chevron';
import { Position, Theme } from '../../common';
import Option from '../../common/Option';

const AccordionItem = ({ id, title, content, onClick, open, icon, theme }) => {
  const iconElement = icon ? cloneElement(icon, { size: 24 }) : null;

  return (
    <div
      id={id}
      className={classNames('np-accordion-item', `np-accordion-item--${theme}`, {
        'np-accordion-item--open': open,
      })}
    >
      <Option
        as="button"
        media={iconElement}
        title={title}
        button={<Chevron orientation={open ? Position.TOP : Position.DOWN} />}
        inverseMediaCircle={false}
        onClick={onClick}
      />
      {open && (
        <div
          className={classNames('np-accordion-item__content', {
            'has-icon': icon,
          })}
        >
          {content}
        </div>
      )}
    </div>
  );
};

AccordionItem.propTypes = {
  content: PropTypes.node.isRequired,
  icon: PropTypes.node,
  id: PropTypes.string,
  open: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  title: PropTypes.node.isRequired,
  theme: PropTypes.oneOf(['light', 'dark']),
};

AccordionItem.defaultProps = {
  icon: null,
  id: null,
  theme: Theme.LIGHT,
};

export default AccordionItem;
