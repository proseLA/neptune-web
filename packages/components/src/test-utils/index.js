import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';

import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 *
 */
function customRender(
  ui,
  { locale = DEFAULT_LOCALE, messages = en, direction, ...renderOptions } = {},
) {
  // eslint-disable-next-line react/prop-types
  var Wrapper = ({ children }) => {
    return (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

/**
 * Custom `renderHook` function which wraps passed elements in Provider component
 * For more info: https://react-hooks-testing-library.com/usage/advanced-hooks#context
 *
 */
function customRenderHook(callback, { locale = DEFAULT_LOCALE, messages = en, direction } = {}) {
  return renderHook(callback, {
    // eslint-disable-next-line react/prop-types
    wrapper: ({ children }) => (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    ),
  });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render, customRenderHook as renderHook };
