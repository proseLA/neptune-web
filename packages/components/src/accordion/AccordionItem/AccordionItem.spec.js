import { Freeze } from '@transferwise/icons';

import { useDirection } from '../../common/hooks';
import { render, userEvent, screen } from '../../test-utils';

import AccordionItem from './AccordionItem';

jest.mock('../../common/hooks');

describe('AccordionItem', () => {
  const props = {
    title: 'This is title number one',
    content: 'Lauri Ipsum has been the industry standard dummy text ever since the 1500s.',
    open: false,
    onClick: jest.fn(),
  };

  beforeAll(() => {
    useDirection.mockImplementation(() => ({ direction: 'ltr', isRTL: false }));
  });

  describe('open / close', () => {
    it('renders an item closed', () => {
      expect(render(<AccordionItem {...props} />).container).toMatchSnapshot();
    });

    it('renders an item open', () => {
      expect(render(<AccordionItem {...props} open />).container).toMatchSnapshot();
    });
  });

  describe('icons', () => {
    it('renders the icon if passed', () => {
      render(<AccordionItem {...props} icon={<span>mock icon</span>} />);

      expect(screen.getByText('mock icon')).toBeInTheDocument();
    });

    it('will always render icons at 24px', () => {
      const { container } = render(<AccordionItem {...props} icon={<Freeze size={16} />} />);

      // eslint-disable-next-line testing-library/no-container
      expect(container.querySelector('.tw-icon-freeze svg')).toHaveAttribute('width', '24');
    });
  });

  describe('onClick', () => {
    it('calls onClick when item is clicked', () => {
      const onClick = jest.fn();

      render(<AccordionItem {...props} onClick={onClick} />);

      userEvent.click(screen.getByRole('button'));

      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
  describe('Right-To-Left', () => {
    it('applies correct css classes when isRTL is true', () => {
      useDirection.mockImplementation(() => ({ direction: 'rtl', isRTL: true }));
      const { container } = render(<AccordionItem {...props} icon={<span>mock icon</span>} open />);
      // eslint-disable-next-line testing-library/no-container
      expect(container.querySelector('.np-accordion-item__content')).toHaveClass('icon-right');
    });
  });
});
