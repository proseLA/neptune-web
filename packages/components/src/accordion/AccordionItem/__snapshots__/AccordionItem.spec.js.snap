// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem open / close renders an item closed 1`] = `
<div>
  <div
    class="np-accordion-item np-accordion-item--light"
  >
    <button
      aria-label="This is title number one"
      class="np-option decision"
    >
      <div
        class="media"
      >
        <div
          class="media-body"
        >
          <div
            class="h5"
          >
            This is title number one
          </div>
        </div>
        <div
          class="media-right"
        >
          <span
            aria-hidden="true"
            class="tw-icon tw-icon-chevron-up tw-chevron chevron-color bottom"
            role="presentation"
          >
            <svg
              fill="currentColor"
              focusable="false"
              height="16"
              width="16"
            >
              <path
                clip-rule="evenodd"
                d="M8 4l-6.6 6.653L2.537 11.8 8 6.293l5.463 5.507 1.137-1.147L8 4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`AccordionItem open / close renders an item open 1`] = `
<div>
  <div
    class="np-accordion-item np-accordion-item--light np-accordion-item--open"
  >
    <button
      aria-label="This is title number one"
      class="np-option decision"
    >
      <div
        class="media"
      >
        <div
          class="media-body"
        >
          <div
            class="h5"
          >
            This is title number one
          </div>
        </div>
        <div
          class="media-right"
        >
          <span
            aria-hidden="true"
            class="tw-icon tw-icon-chevron-up tw-chevron chevron-color top"
            role="presentation"
          >
            <svg
              fill="currentColor"
              focusable="false"
              height="16"
              width="16"
            >
              <path
                clip-rule="evenodd"
                d="M8 4l-6.6 6.653L2.537 11.8 8 6.293l5.463 5.507 1.137-1.147L8 4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </button>
    <div
      class="np-accordion-item__content"
    >
      Lauri Ipsum has been the industry standard dummy text ever since the 1500s.
    </div>
  </div>
</div>
`;
